### ISSUE TITLE
[InstCombine] failed to fold fptoui(fdiv(uitofp(x),c)) into udiv(x, c)

### ISSUE THREAD

**Issue by Andarwinux** [llvm/llvm-project#205305]
https://godbolt.org/z/E84Y45714
```c
unsigned int a(unsigned int x, unsigned int y){
    return (double)x/(double)y;
}

unsigned int b(unsigned int x){
    return a(x, 7);
}

unsigned int c(unsigned int x){
    return x / 7;
}
```
```asm
a:
        vcvtusi2sd      xmm0, xmm15, edi
        vcvtusi2sd      xmm1, xmm15, esi
        vdivsd  xmm0, xmm0, xmm1
        vcvttsd2usi     eax, xmm0
        ret

.LCPI1_0:
        .quad   0x401c000000000000
b:
        vcvtusi2sd      xmm0, xmm15, edi
        vdivsd  xmm0, xmm0, qword ptr [rip + .LCPI1_0]
        vcvttsd2usi     eax, xmm0
        ret

c:
        mov     edx, edi
        movabs  rax, 2635249153617166336
        mulx    rax, rax, rax
        ret
```
In the past, since x86 lacked vector udiv instructions, people typically used fdiv to replace udiv. However, after #205263, x86 backend will be able to use vector fdiv to lower vector udiv, so this is no longer necessary.

But existing code won't switch back to udiv. The drawback of manually using fdiv to replace udiv is the backend will no longer be able to apply optimizations when the divisor is known. InstCombine should detect this pattern and convert it back to udiv.

**Comment by Andarwinux** [comment]
This may also be useful for aarch64, as sve has native vector udiv. In this case, folding can be applied unconditionally even if the divisor not a known constant.

**Comment by RKSimon** [comment]
> This may also be useful for aarch64, as sve has native vector udiv. In this case, folding can be applied unconditionally even if the divisor not a known constant.

If we're doing this in instcombine then the target is irrelevant - if we're limiting it vectors then a cost driven fold in vectorcombine is an alternative

**Comment by Andarwinux** [comment]
> > This may also be useful for aarch64, as sve has native vector udiv. In this case, folding can be applied unconditionally even if the divisor not a known constant.
> 
> If we're doing this in instcombine then the target is irrelevant - if we're limiting it vectors then a cost driven fold in vectorcombine is an alternative

Created #205502.

**Comment by goranmoomin** [comment]
I wanted to try getting this implemented myself, sorry if this is more of a nuisance.

My understanding is that implementing this would mean that the `src` function below be optimized to `tgt` in InstCombine?
(I probably should implement this in `InstCombinerImpl::visitFPToUI` as a `foldFPToUIOfFDiv` or something like that.)

```llvm
define i32 @src(i32 %x) {
  %1 = uitofp i32 %x to double
  %2 = fdiv double %1, 7.000000e+00
  %3 = fptoui double %2 to i32
  ret i32 %3
}

define i32 @tgt(i32 %x) {
  %1 = udiv i32 %x, 7
  ret i32 %1
}
```

**Comment by goranmoomin** [comment]
I've been thinking on when the transformation `fptoui(fdiv(uitofp(X), C)) -> udiv(X, C)` would be sound,
and for the constant case I think it would be valid if
- C > 0 and
- FP's precision (mantissa width) is geq than X's bit width.

Proof:

let X be an N-bit uint, C a uint constant, 1 <= C < 2^N
the FP type of fptoui/fdiv/uitofp has precision p (mantissa width) with N <= p

then, `uitofp X` and C is both exactly representable in the FP type, hence `fptoui(fdiv(uitofp(X), C)) = floor(rne(X/C))`.
now we show that `floor(rne(X/C)) = floor(X/C)` by:

let `H(m) = (m-pred(m))/2 = ulp(pred(m))/2`  (where m is a uint >= 2 represented in p-bit fp):
then, `H(1)=2^(-p-1)`, and for all m >= 2, `H(m) < (m-1)/2^p`:
- proof: for m==1, trivial.
- let m >= 2, let `e = floor(log2(m))`,
- then if `2^e < m < 2^(e+1)`, then `ulp=2^(e-p+1)`, so `H(m) = ulp/2 = 2^(e-p) <= (m-1)/2^p`
- else, `m == 2^e`, then `ulp=2^(e-p)`, so `H(m) = ulp/2 = 2^(e-p-1) <= (m-1)/2^p`

now, let `y = X/C = q + r/C`, and `m = q+1`. if r == 0: trivial, so let's only consider r >= 1.

`rne(y) >= q` is trivial since `q < 2^N <= 2^p` and therefore exactly representable in p-bit fp, and y >= q therefore rne(y) >= q

`rne(y) < m` is equivalent to `y < m - H(m)` is equivalent to `H(m) < m - y`, and `m-y = (C-r)/C >= 1/C > q/2^p = (m-1)/2^p > H(m)`.

therefore, `floor(rne(y)) == q`, hence the proof is complete.

I think this is correct.

**Comment by Andarwinux** [comment]
Thanks, that sounds very reasonable.

**Comment by goranmoomin** [comment]
I have an initial implementation for the constant case, but then I also think this case generalizes to the variable divisor case, and also for signed cases as well. I'd assume that if it's possible, the variable divisor case should also be implemented? I'm not sure if the signed case would also be beneficial though.

**Comment by Andarwinux** [comment]
> I have an initial implementation for the constant case, but then I also think this case generalizes to the variable divisor case, and also for signed cases as well. I'd assume that if it's possible, the variable divisor case should also be implemented? I'm not sure if the signed case would also be beneficial though.

The profitability of variable divisor folding is highly dependent on the target. Folding variable divisor on x86 is never profitable (even for scalar, udiv is not necessarily faster than uitofp+fdiv+fptoui, see https://github.com/llvm/llvm-project/issues/201319), while on aarch64 it's only profitable with sve in the case of vXi64/i32 (sve has native i64/i32 vector udiv, but not i16/i8).

as RKSimon said, this should be placed in VectorCombine or later.

It should also make sense to implement the signed case. According to [LangRef](https://llvm.org/docs/LangRef.html#i-sdiv), INT_MIN/-1 also leads to undefined behavior:

>The value produced is the signed integer quotient of the two operands rounded towards zero.
>Note that signed integer division and unsigned integer division are distinct operations; for unsigned integer division, use ‘udiv’.
>Division by zero is undefined behavior. For vectors, if any element of the divisor is zero, the operation has undefined >behavior. Overflow also leads to undefined behavior; this is a rare case, but can occur, for example, by doing a 32-bit >division of -2147483648 by -1.
>If the exact keyword is present, the result value of the sdiv is a [poison value](https://llvm.org/docs/LangRef.html#poisonvalues) if the result would be rounded.

So there's no need to worry that folding signed cases will break expected behavior, but I'm not a llvm developer - this should be confirmed by InstCombine maintainers.

**Comment by goranmoomin** [comment]
My understanding is that the signed case would be valid except for the case of INT_MIN / -1, since sitofp+fdiv+fptosi would only produce a poison value (when fptosi overflows), while the sdiv would be undefined behavior. Which would mean that the transformation wouldn't be sound for that case.

Though it seems like InstCombine already canonicalizes fdiv X, -1 into a fneg, so it wouldn't really matter handling the C = -1 case.

**Comment by Andarwinux** [comment]
> My understanding is that the signed case would be valid except for the case of INT_MIN / -1, since sitofp+fdiv+fptosi would only produce a poison value (when fptosi overflows), while the sdiv would be undefined behavior. Which would mean that the transformation wouldn't be sound for that case.

oops, you're right. Although the user's intent is just signed integer division, but folding sitofp+fdiv+fptosi into sdiv at IR level actually introduces undefined behavior that wasn't there before.

> Though it seems like InstCombine already canonicalizes fdiv X, -1 into a fneg, so it wouldn't really matter handling the C = -1 case.

https://godbolt.org/z/E84Y45714

That's interesting, in the case of -1, it's already been fully folded.

**Comment by goranmoomin** [comment]
For the proof on the soundness of the transformation for both signed and unsigned cases, see: https://goranmoomin.github.io/llvm-project--205305.html.

Note: producing the typst/html & md document was LLM-assisted. The proof itself was done and reviewed by me.

<details>
<summary>Proof in Markdown</summary>

## Soundness of `fdiv`/`fptoi` to integer division

Consider the LLVM fold, lane-wise for vectors,

```
fpto[u/s]i (fdiv ([u/s]itofp X to F), C_fp) to iN
  --> [u/s]div iN X, C
```

where `F` is a binary floating-point type with precision $p$ significand bits (including the hidden bit), and `C_fp` is exactly the FP value of the integer constant `C`.

### Conditions

For the unsigned form:

- $C > 0$
- $N \leq p$

For the signed form:

- $C \neq 0$
- $N - 1 \leq p$
- If $C = -1$, the fold to `sdiv` is sound only when $X \neq \text{INT\\_MIN}$ is known. Otherwise `sdiv INT_MIN, -1` is immediate undefined behavior, while the FP path only produces poison at `fptosi`.

Equivalently, the signed constant cases are:

```
C > 0   : sound if N - 1 <= p
C < -1  : sound if N - 1 <= p
C == -1 : sound only if X != INT_MIN is known
C == 0  : not sound as division
```

The precision bounds ensure that the integer inputs to the FP division are exact: all integers with magnitude at most $2^p$ are exactly representable.

### Rounding lemma

Let $R_p$ be round-to-nearest-even to precision $p$. If $0 \leq A \leq 2^p$, $1 \leq B \leq 2^p$, and $q = \lfloor A/B \rfloor$, then

$$\mathrm{trunc}\left(R_p\left(\frac{A}{B}\right)\right) = q.$$

**Proof.** Write $A = qB + r$, where $0 \leq r < B$. If $r = 0$, then $A/B = q$ exactly, and $q \leq 2^p$, so rounding preserves it.

Assume $r > 0$ and let $m = q + 1$. The quotient lies in $[q, m)$, so it cannot round below $q$. It remains to show it cannot round up to $m$. Let $H(m)$ be half the gap from $m$ to the previous representable FP number. For $q > 0$,

$$H(m) \leq \frac{q}{2^p}.$$

Since $qB < A \leq 2^p$, we have $B < \dfrac{2^p}{q}$, hence

$$\frac{1}{B} > \frac{q}{2^p} \geq H(m).$$

Also,

$$m - \frac{A}{B} = \frac{B - r}{B} \geq \frac{1}{B} > H(m).$$

Thus $A/B$ is strictly below the rounding threshold for $m$, so $R_p(A/B) < m = q + 1$, and truncation gives $q$.

If $q = 0$, then $m = 1$ and $H(1) = 2^{-p-1}$, while $B \leq 2^p$ gives $\dfrac{1}{B} \geq 2^{-p} > H(1)$, so the same conclusion holds.

### Signed case

Let `X` have signed value $a$ and divisor $c = C$. Define

$$A = |a|,\quad B = |c|,\quad s = \mathrm{sign}(a)\mathrm{sign}(c).$$

Then $a/c = s(A/B)$. Under $N - 1 \leq p$, both $A$ and $B$ are at most $2^{N-1} \leq 2^p$, so `sitofp` and the FP constant are exact. By the lemma,

$$R_p\left(\frac{A}{B}\right) \in [q,\, q + 1),\quad q = \left\lfloor \frac{A}{B} \right\rfloor.$$

Round-to-nearest-even is symmetric, so $R_p(-z) = -R_p(z)$. Therefore the FP path produces $s\,R_p(A/B)$ before `fptosi`. Truncation toward zero gives

$$\mathrm{fptosi}\left(s\,R_p\left(\frac{A}{B}\right)\right) = sq = \mathrm{sdiv}(a,\, c),$$

provided the integer division is defined. The only signed overflow case is $\text{INT\\_MIN} / {-1}$, which is why `C == -1` needs the extra proof obligation.

The unsigned proof is the same lemma with $A = X$, $B = C$, and $N \leq p$.

</details>
